### ISSUE TITLE
Hardware requirements.

### ISSUE THREAD

**Issue by FatemaD1577** [ultralytics/ultralytics#4106]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to use Yolo V8 to train custom dataset for object detection task on my system. What are the hardware requirements of CPU and GPU?

**Comment by xaristeidou** [comment]
@FatemaD1577 It really depends on a lot of parameters to answer this question. Here are some examples:

- It depends on the model you will be using, for example `yolov8n` is much smaller model that `yolov8x` and will consume far less GB of GPU or RAM space for the training. 
- Also, the size of the training images which is specified in `imgsz` parameter is very crucial, larger image size for training images will consume more GB while training.
- If your GPU is any new nvidia model and supports Mixed Precision then less space will be needed for the training. 
- Batch size is also a crucial parameter which plays a significant role in the training procedure and higher `batch_size`  consume more memory while training.


Typically if you have a 4GB GPU without support for Mixed Precision, you will need to train using `imgsz = 320` which i think is the lowest resolution accepted by the model and small `batch_size` like `1,2,4`. 

If you have 8GB of GPU or more, then you have nothing to worry you can train easy.

On CPU your training is going to be slower but if you have 12GB or more of RAM you can train.

**Comment by glenn-jocher** [comment]
@xaristeidou, the hardware requirements for training with YOLOv8 on your system largely depend on several key factors:

**Model Size**: Different models within YOLOv8, such as `yolov8n` (a smaller model) and `yolov8x` (a larger model), have different resource requirements. Bigger models consume more memory but are typically more accurate.

**Image Size (`imgsz` parameter)**: Larger image sizes mean more pixel data to process, which can significantly increase memory usage. You'll have to strike a balance between the level of detail you want and the memory you have available.

**Mixed Precision**: If your GPU, especially a recent Nvidia model, supports Mixed Precision, it can reduce memory usage and speed up training without compromising the resulting model's performance.

**Batch Size**: The batch size refers to the number of training instances used simultaneously during a single iteration of model training. Large batch sizes require more memory but can lead to faster, more stable training.

For a 4GB GPU without Mixed Precision support, you could consider training with `imgsz = 320` (relatively low resolution) and a small batch size (e.g., `1,2,4`). This setup should enable training without exceeding your hardware's capabilities.

With an 8GB GPU or larger, you should have more flexibility in adjusting these parameters to optimize performance and accuracy.

CPU-based training is generally slower than GPU-based training. However, with 12GB or more of RAM, you should be able to train your custom dataset reasonably effectively. Please remember, these are all estimations, and the actual requirements may vary depending on several factors including the complexity of your dataset.

**Comment by FatemaD1577** [comment]
Hi @xaristeidou & @glenn-jocher, Thanks for the suggestions. 

Please find below details related to data:
Images: Black and white
Number of images: 5k

Training parameters preferred:
imgsz = 640 
batch size = 32 

Regarding the models, right now we have worked on colab and have been using yolov8n/s models. We are planning to set up a dedicated GPU machine for training. Kindly guide me on the recommended model in that case, the RAM, memory & GPU requirement along with preferred imgsz and batch size for efficient performance.

**Comment by glenn-jocher** [comment]
@FatemaD1577 hello, thanks for providing the dataset details and your preferred training parameters.

In terms of model selection, choosing between 'yolov8n' or 'yolov8s' will depend on your specific needs for speed and accuracy. Typically, the 'n' model is smaller and faster but slightly less accurate, while the 's' model is slightly larger, slower, but may yield better detections.

For the image size (`imgsz`), if your GPU has enough memory, you should be able to use the preferred size of 640. However, if you face memory-related errors, you might need to reduce it.

For the batch size, 32 should be fine. Larger batch sizes typically require more GPU memory, but they can also make the training more stable and faster.

In terms of hardware requirements, these would be the recommendations for decent performance:

GPU Memory: Ideally, you should go for a GPU with at least 8GB memory. However, a 12 GB GPU memory would be more comfortable for your preferred training parameters.
RAM: A minimum of 16 GB RAM is suggested, although 32 GB would provide a more efficient training environment.
CPU Cores: A quad-core CPU (4 cores) should be sufficient. More cores might benefit multi-threaded tasks but usually does not affect model training significantly.
OS: A 64-bit operating system that supports your selected hardware and necessary software (like CUDA for Nvidia GPUs) should work. Common choices include various distributions of Linux, such as Ubuntu.

Remember, these are general suggestions. Specific hardware requirements can depend on various factors, including the individual complexity and characteristics of your dataset. You should monitor your system's utilization during training to ensure it's being used efficiently.

**Comment by FatemaD1577** [comment]
@glenn-jocher Thanks for the suggestions. This seems to help.

The parameters I have mentioned are what we are using currently, to train on colab due to the memory and time constraints for the GPU allocation. Should we go for a different model like l/x, and accordingly any other batch size preferred for better model performance. Also, the corresponding changes in GPU memory and ram requirements.

I had come across one issue of yolo V5 regarding image size. There it was mentioned that default size for yolo V5 P5 is 640 and for yolo V5 P6 is 1028. If you could kindly guide me with the required sizes for different yolo V8 models as well it will be a great help.

With image size in mind, I have a question as well. Does how image size affect the quality of training and memory and RAM requirements?

**Comment by glenn-jocher** [comment]
Hi @FatemaD1577, 

Glad to be of help. The selection of the model size ('n', 's', 'l', 'x', etc.) indeed depends on your specific requirements for speed vs accuracy trade-off. The 'l' and 'x' models are indeed larger and usually provide better accuracy but consume more memory and take more time to train compared to 'n' and 's' models. The increase in model size also typically results in elevation in GPU memory and RAM requirements.

Regarding image sizes, YOLOv8 also works on a similar principle. For 'l' and 'x' models, as the model size increases, generally a larger image input size is recommended for better model performance, considering that the hardware resources like GPU memory and RAM are available. But take note that increasing image size also escalates the memory used during training, and can also influence training time based on how large the size has been set.

As for the cloud deployments, choosing between 'l'/'x' models or smaller ones like 'n'/'s' tends to be a matter of balancing the performance and resource cost. Larger models like 'l'/'x' can potentially deliver superior accuracy, but they also demand more computational resources, which may be expensive in a cloud-based setting.

The image size directly impacts the quality of training and memory and RAM requirements. Larger image sizes typically enable the model to learn more detailed features, potentially improving the model's performance. However, this also increases the GPU and RAM requirements as larger images demand more memory to process. It is recommendable to choose an image size that balances memory requirement and training quality, considering the available hardware resources.

Hope this help further clarifies your doubts.

**Comment by FatemaD1577** [comment]
Thank you, @glenn-jocher !! Will start working with the requirements and keep on updating as and when required.

**Comment by glenn-jocher** [comment]
Hello @FatemaD1577,

I'm glad to hear that you found the provided information helpful. I look forward to hearing about your progress with implementing your YOLOv8 model. Remember, the requirements outlined in our discussion will provide a good starting point.

As you work through training, don't hesitate to adjust your parameters if you find they improve your model's optimization and performance. This is often an iterative process of fine-tuning to achieve the best balance between accuracy, speed, and resource usage. 

Feel free to come back with questions or updates. I'm here to assist you through this! Happy coding!

**Comment by luongnguyen10** [comment]
is there any way to calculate the compute unit, memory to choose vga, cpu and ram from model size, input image?

**Comment by glenn-jocher** [comment]
@luongnguyen10 hello there,

Certainly, there is a way to estimate the computational resources necessary for training YOLOv8 models, which will allow you to choose appropriate hardware for the task. It all depends on the size of the model you're planning on using, and the input image resolution.

For model size, larger models like 'l' and 'x' variants will require more GPU memory, primarily due to the increased number of parameters they contain. If you plan on using these larger models, it's recommended to consider a GPU with a higher amount of memory, perhaps starting at 12GB or more. Smaller models such as 'n' and 's' should comfortably run on GPUs with around 8GB of memory.

The input image resolution (`imgsz`) also plays a significant role. Larger images will require more memory for both the GPU (for model training) and the system RAM (for data loading and preprocessing). Increasing the resolution will allow the model to learn finer details but at the cost of increased computational demand. For inference, the GPU's raw compute power (measured in TFLOPS, or tera floating-point operations per second) will also impact the speed at which the model can process images.

To determine the necessary CPU and system RAM, remember that the CPU is generally less critical for the direct model training process, which is primarily handled by the GPU. However, a faster CPU can help in pre-processing data and in handling the data pipeline more efficiently. Regarding RAM, a general rule is to have at least twice as much RAM as GPU memory to ensure smooth operation, especially when working with large datasets or high-resolution images.

In summary, for larger models and higher image resolutions, invest in higher GPU memory and consider a GPU with high compute capabilities, accompanied by sufficient system RAM to ensure smooth data handling. Smaller models and lower resolutions provide more flexibility in terms of hardware choice. Remember that these are general guidelines, and the specific requirements can vary based on the exact model complexity and dataset characteristics. 

Best of luck with your YOLOv8 implementations, and I hope this helps you plan your hardware acquisition!

**Comment by Faheemibrahim** [comment]
i have and 4GB GTX Nvidia 1050 and my laptop freezes while training. is this hardware limitations ? or can i add other parameters to avoid the laptop from freezing. I am running the code on visual studio on ubuntu 20,04 and the code below is what i am using for training. i have tried cache=true and it was the same issue. i have played around with the batch size i haven't tired training the model on cpu for long.   

from ultralytics import YOLO

model = YOLO('yolov8n.pt')

results = model.train(data='/home/faheem/VENVS/cv/pythoncv/Rock-Paper-Scissors-SXSW-11/data.yaml',batch = 1, epochs=10, imgsz=320,device = '0')

**Comment by elcodered** [comment]
Hello, i have a problem installing ultralytics. I am using Raspberry Pi 3 Model B, with an operating system of Raspios Debian version: 11 (bullseye) 32-bit, Kernel Version: 5.10

It started from preparing metadata (pyproject.toml) started
It started from preparing metadata (pyproject.toml) finished with status 'error'

Preparing metadata (pyproject.toml) did not run successfully. And so much more errors below.

Should I upgrade to a higher Version and run 64-bit operating system of Raspberry Pi OS? Please beep me back @glenn-jocher ASAP. Thanks

**Comment by glenn-jocher** [comment]
@elcodered hey there! 👋 It seems like you're running into some installation issues on your Raspberry Pi 3. While upgrading to a 64-bit OS could potentially help, it might not be necessary just yet. Could you try installing the package in a virtual environment? Here's a quick way to set one up:

```bash
python3 -m venv ultralytics-env
source ultralytics-env/bin/activate
pip install --upgrade pip
pip install ultralytics
```

This ensures you're working with a clean slate and the latest version of pip. If you still encounter errors, please share the complete error messages. That way, I can offer more targeted assistance. Thanks for reaching out! 🛠️

**Comment by AlejandroDiazD** [comment]
Hello there. Are there any minimum hw requirements for inference with yolov8? In my case I'm trying to run yolov8n on an edge device that just have 400mb RAM available, but it seems to not be enough. In addition to that minimum hw requirements (that in the case of RAM empirically seems to be about that 400MB), is there anyway to reduce the required RAM when loading the model?
Thank you in advance.

**Comment by glenn-jocher** [comment]
@AlejandroDiazD hello! 🌟 Indeed, running YOLOv8 models, especially on edge devices, does have some hardware requirements. Typically, 400MB of RAM might be on the lower end, particularly for loading and running neural networks as complex as YOLOv8.

Reducing RAM usage for model loading isn't straightforward due to the inherent size of the model architecture and weights. However, you can try a few things:

1. **Optimize the model**: Use tools like ONNX or TensorFlow Lite to convert and potentially optimize the model, which might reduce its load size.
   
2. **Model Variant**: Ensure you're using the smallest version of YOLOv8, i.e., yolov8n, for minimal resource consumption.

3. **Split Loading**: If your application permits, consider loading parts of the model on demand rather than keeping the entire model in memory.

Remember, these methods might reduce RAM usage modestly but won't drastically change the hardware requirements. Upgrading your edge device or selecting a more capable one might be necessary for smoother operations. 

Good luck with your project! 🚀

**Comment by Abdul-lah-prog** [comment]
i'm working on a smart traffic monitoring device. Basically i'm struggling in choosing a hardware that can easily run YOLO algorithms like v8 . Hoping to get some suggestions.

**Comment by pderrenger** [comment]
@Abdul-lah-prog for running YOLOv8 on a smart traffic monitoring device, consider hardware with a CUDA-compatible GPU for optimal performance. Devices like NVIDIA Jetson Nano or Xavier are popular choices for edge AI applications due to their balance of power and efficiency. Ensure the device has at least 4GB of RAM for smoother operation. If you're constrained by budget or power, you might also explore Coral Edge TPU, which can handle lighter models effectively. Always check the latest compatibility and performance benchmarks for your specific use case.

**Comment by subbiah-cape** [comment]
Hi @glenn-jocher, I've been using YOLO V8X-seg and would like to know the minimum hardware requirements for running the model on a CPU-based system. I've deployed the model in Azure Kubernetes Service (AKS) and aim to handle at least 350 requests per minute.

Currently, I'm using a Standard_D8ds_v5 VM with 3 nodes and 6 pods, where each pod is configured with a minimum of 2 vCPUs and a maximum of 5 vCPUs. During load testing, the requests are being processed, but it takes around 4 minutes to complete 300 requests per minute.

Could you please help me reduce the latency and suggest a more suitable VM configuration in AKS to achieve this goal?

**Comment by glenn-jocher** [comment]
To reduce latency and improve performance, consider optimizing your model by using a smaller variant like YOLOv8n-seg if feasible, or explore model quantization techniques to speed up inference. Additionally, increasing the number of vCPUs per pod or scaling up the number of nodes in your AKS cluster could help distribute the load more effectively.

**Comment by subbiah-cape** [comment]
Hi Glen,

I've recently switched to a GPU-based system and converted our custom-trained YOLOv8x-seg model to TensorRT. The TensorRT model runs well on the development server (Standard_NC8as_T4_v3).

However, when I dockerize the model in the same dev server for inference, it fails to produce predictions, while the best.pt model works fine and achieves a 0.93 mAP. Additionally, I'm encountering an error when running image predictions with the TensorRT model.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/miniconda/lib/python3.10/site-packages/azureml_inference_server_http/server/user_script.py", line 132, in invoke_run
    run_output = self._wrapped_user_run(**run_parameters, request_headers=dict(request.headers))
  File "/opt/miniconda/lib/python3.10/site-packages/azureml_inference_server_http/server/user_script.py", line 156, in <lambda>
    self._wrapped_user_run = lambda request_headers, **kwargs: self._user_run(**kwargs)
  File "/var/azureml-app/inference-code/score.py", line 126, in run
    results_img = model(image, conf=0.65)
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/model.py", line 176, in __call__
    return self.predict(source, stream, **kwargs)
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/model.py", line 555, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
  File "/opt/miniconda/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 36, in generator_context
    response = gen.send(None)
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 274, in stream_inference
    s[i] += self.write_results(i, Path(paths[i]), im, s)
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 334, in write_results
    string += f"{result.verbose()}{result.speed['inference']:.1f}ms"
  File "/opt/miniconda/lib/python3.10/site-packages/ultralytics/engine/results.py", line 663, in verbose
    log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
KeyError: 17

**Comment by subbiah-cape** [comment]
I need to deploy the TensorRT model in AzureML. After resolving multiple issues, I successfully trained the model, converted it to TensorRT in AzureML, registered, and deployed it. However, when I attempt inference with the TensorRT model, I encounter the same issue as previously mentioned, where predictions are not generated.

In AzureML, I'm using the same server configuration for both training and inference. During training, I used the AzureML driver image (FROM mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.8-cudnn8-ubuntu22.04:latest). Below are the Docker file requirements:

`FROM mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.8-cudnn8-ubuntu22.04:latest

ENV DEBIAN_FRONTEND noninteractive

RUN apt update
RUN ACCEPT_EULA=Y apt-get upgrade -y
RUN TZ=Etc/UTC apt install -y tzdata
RUN apt install --no-install-recommends -y gcc git zip curl htop libgl1-mesa-glx libglib2.0-0 libpython3-dev gnupg g++
RUN apt upgrade --no-install-recommends -y openssl tar

# Install specific versions of Python packages
RUN pip install azureml-inference-server-http
RUN pip install torch
RUN pip install ultralytics==8.2.100
RUN pip install mlflow
RUN pip install azureml-contrib-services`

Are there any additional packages that need to be added here?

**Comment by glenn-jocher** [comment]
It seems like the issue might be related to missing dependencies specific to TensorRT. Ensure that the TensorRT runtime is installed in your Docker image. You can add it by including `tensorrt` in your Dockerfile. Additionally, verify that your model's class labels are correctly mapped, as the error suggests a potential mismatch.

**Comment by subbiah-cape** [comment]
The model's class labels are correctly mapped, and it was working fine in the development environment. However, after dockerizing the model and running inference, the above error appears. All other packages and drivers are identical to those in the development environment.

**Comment by subbiah-cape** [comment]
I trained the model with 4 classes, and when I check the class names, they appear correctly. However, an error occurs when I try to predict an image.

`>>> emodel.names
{0: 'Moderatefit', 1: 'Poorfit', 2: 'Fairfit', 3: 'Goodfit'}
>>> emodel.predict('/app/0a0a0420-b43f-11eb-8039-6bc8d7ec8b40-checkinloweroff-1637197736484.jpeg',conf=0.50,task='segment')

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/venv/lib/python3.10/site-packages/ultralytics/engine/model.py", line 555, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "/opt/venv/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
  File "/opt/venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 36, in generator_context
    response = gen.send(None)
  File "/opt/venv/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 274, in stream_inference
    s[i] += self.write_results(i, Path(paths[i]), im, s)
  File "/opt/venv/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 334, in write_results
    string += f"{result.verbose()}{result.speed['inference']:.1f}ms"
  File "/opt/venv/lib/python3.10/site-packages/ultralytics/engine/results.py", line 663, in verbose
    log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
KeyError: 9`

**Comment by glenn-jocher** [comment]
It seems like there might be an issue with the class index mapping during inference. Please ensure that the model's class indices align with the expected range. Double-check your dataset and model configuration to verify that all class indices are correctly defined and consistent. If the issue persists, consider retraining the model to ensure proper class mapping.

**Comment by subbiah-cape** [comment]
The trained model was saved at runs/segment/train47/weights/best.engine, and it works correctly when I use this path. However, if I copy best.engine to a different folder, the model fails to make predictions.

`CUDA available: True
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Loading /data/mybyte/subbiah/github/MyByte-ML_Model_Training/docker_new/inference_cuda_12_4/train47/weights/bestcopy.engine for TensorRT inference...
[11/05/2024-12:56:12] [TRT] [I] Loaded engine size: 74 MiB
[11/05/2024-12:56:12] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +583, now: CPU 0, GPU 701 (MiB)

WARNING ⚠️ NMS time limit 2.050s exceeded
Traceback (most recent call last):
  File "/data/mybyte/subbiah/github/MyByte-ML_Model_Training/onnx_load_test.py", line 44, in <module>
    rt = eng_model(f'{image_path}/{img}')
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/model.py", line 176, in __call__
    return self.predict(source, stream, **kwargs)
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/model.py", line 555, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 36, in generator_context
    response = gen.send(None)
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 274, in stream_inference
    s[i] += self.write_results(i, Path(paths[i]), im, s)
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/predictor.py", line 334, in write_results
    string += f"{result.verbose()}{result.speed['inference']:.1f}ms"
  File "/data/mybyte/subbiah/tensorrt_env/lib/python3.10/site-packages/ultralytics/engine/results.py", line 663, in verbose
    log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
KeyError: 9`

**Comment by glenn-jocher** [comment]
@subbiah-cape it seems like there might be an issue with the model's environment or dependencies when moving the `.engine` file. Ensure that all necessary files and configurations are correctly set up in the new location, and verify that the model's task is explicitly defined to avoid assumptions. If the problem persists, consider checking the TensorRT setup and any related paths or permissions.

**Comment by Y-T-G** [comment]
@subbiah-cape You have to specify `task` while loading the TensorRT model. It says that in the warning.

> WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.

`model = YOLO("best.engine", task="segment")`

**Comment by subbiah-cape** [comment]
Does the model's performance decrease when exported from PyTorch to TensorRT?

For best.pt, I achieved a mean average precision of 0.92. Will this remain the same after converting to TensorRT?

**Comment by Y-T-G** [comment]
@subbiah-cape It would be similar if you don't use `int8`

**Comment by subbiah-cape** [comment]
Thanks for the details, @Y-T-G @glenn-jocher . It seems that using int8 reduces performance, so when exporting with standard quantization using model.export(format="engine"), the model's speed is similar to the PyTorch model. Are there any specific parameters that could enhance the speed beyond the PyTorch model's without compromising performance?

**Comment by Y-T-G** [comment]
I doubt it's similar to PyTorch performance. TensorRT is at least twice or faster than PyTorch.

**Comment by subbiah-cape** [comment]
I was using a T4 GPU instance, and I noticed that using parameters like half=True during export improves performance. However, without any additional arguments, the speed remains about the same as pytorch

**Comment by subbiah-cape** [comment]
Here’s the code I used to measure the speed.
eng_model = YOLO('/data/mybyte/subbiah/github/MyByte-ML_Model_Training/runs/segment/train47/weights/best.engine')

`image_path= '/data/mybyte/subbiah/github/MyByte-ML_Model_Training/final_fullteeth_NP/test'
all_time=[]
loop_start_time_gpu = time.time()
for img in os.listdir (image_path):
    # all_time=[]
    start_time_gpu = time.time()
    rt = eng_model(f'{image_path}/{img}')
    speed=rt[0].speed
    total_time = speed['preprocess'] + speed['inference'] + speed['postprocess']
    end_time_gpu = time.time()
    ontime_gpu_ms = (end_time_gpu - start_time_gpu) * 1000
    all_time.append(total_time)
    print(f"Execution time for TRT model (GPU): {ontime_gpu_ms:.2f} ms")
loop_end_time_gpu = time.time()
loop_ontime_gpu_ms = (loop_end_time_gpu - loop_start_time_gpu) * 1000
print(f"Execution loop time for TRT model (GPU): {loop_ontime_gpu_ms:.2f} ms")

all_time_f=all_time[1:-1]
print(f'max time to predict image {np.max(all_time_f)}ms')
print(f'min time to predict image {np.min(all_time)}ms')
sumtime = np.sum(all_time)
print(f'average time to predict image {sumtime/100}ms')
`

**Comment by Y-T-G** [comment]
The first inference is going to be slower because it sets up TensorRT. You should be excluding that. And if you're inferring in images, some of the time is spent on loading the images from the disk which is a lot slower than inference.

**Comment by Y-T-G** [comment]
And half=True should have pretty similar accuracy to PyTorch

**Comment by subbiah-cape** [comment]
I observed the speed of the first inference and recorded the  inference  from each output .  
`    rt = eng_model(f'{image_path}/{img}')
    speed=rt[0].speed
    total_time = speed['preprocess'] + speed['inference'] + speed['postprocess']`
 believe that loading images from the disk likely doesn't make much of a difference.

**Comment by subbiah-cape** [comment]
`image 1/1 /data/mybyte/subbiah/github/MyByte-ML_Model_Training/final_fullteeth_NP/test/781b9130-7c94-11ed-88e9-1f3b5dda8ba6-checkinloweron-1691115022362.jpeg: 640x640 1 Fairfit, 13.2ms
Speed: 1.7ms preprocess, 13.2ms inference, 1.8ms postprocess per image at shape (1, 3, 640, 640)
Execution time for TRT model (GPU): 26.99 ms`

**Comment by Y-T-G** [comment]
> believe that loading images from the disk likely doesn't make much of a difference.

It takes more time than the inference.
